### ISSUE TITLE
one node gives many  Exceptional future ignored: seastar::named_semaphor e_timed_out (Semaphore timed out: _read_concurrency_sem),

### ISSUE THREAD

**Issue by pidiaquez** [scylladb/scylla#6817]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash):  4.0.0-0.20200505.d95aa77b62
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): centos7

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker):
Hardware: sockets= cores= hyperthreading= memory=
Disks: (SSD/HDD, count)

error seen is

Exceptional future ignored: seastar::named_semaphor e_timed_out (Semaphore timed out: _read_concurrency_sem),

logs at

e9a6a11a-1502-4b9e-9bfd-043098f2e54f

**Comment by pidiaquez** [comment]
Datacenter: AWS_US_EAST_1
=========================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
DN  10.250.2.11  44.67 GB   256          ?       hhhhhhhhhhhhh  us-east-1c
UN  10.250.1.11  159.7 GB   256          ?       iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii  us-east-1b
UN  10.250.0.11  159.6 GB   256          ?       jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj  us-east-1a

[support@ip-10-250-1-11 ~]$ nodetool status
Datacenter: AWS_US_EAST_1
=========================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
DN  10.250.2.11  44.67 GB   256          ?       hhhhhhhhhhhhhh  us-east-1c
UN  10.250.1.11  159.86 GB  256          ?       iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii us-east-1b
UN  10.250.0.11  159.71 GB  256          ?       jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj  us-east-1a

[support@ip-10-250-2-11 ~]$ nodetool status
Datacenter: AWS_US_EAST_1
=========================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load      Tokens      Owns    Host ID                              Rack
UN  10.250.2.11  44.67 GB  256          ?      hhhhhhhhhhhhhhh  us-east-1c
DN  10.250.1.11  46.53 GB  256          ?      iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii  us-east-1b
DN  10.250.0.11  46.62 GB  256          ?      jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj  us-east-1a

**Comment by pidiaquez** [comment]
node affected is 10.250.2.11, issue occur while  load of data was in progress. 

`Jul 10 00:43:20 ip-10-250-2-11 scylla[1804]:  [shard 2] seastar - Exceptional future ignored: seastar::named_semaphore_timed_out (Semaphore timed out: _read_concurrency_sem), backtrace:   0x331b31d
                                                            0x2e30f07
                                                            0x19cc19f
                                                            0x2e7ac30
                                                            0x2e7ae3f
                                                            0x2eb2df5
                                                            0x2ec141c
                                                            0x2e4caed
                                                            /opt/scylladb/libreloc/libpthread.so.0+0x94e1
                                                            /opt/scylladb/libreloc/libc.so.6+0x1016a2
Jul 10 00:43:20 ip-10-250-2-11 scylla[1804]:  [shard 2] seastar - Exceptional future ignored: seastar::named_semaphore_timed_out (Semaphore timed out: _read_concurrency_sem), backtrace:   0x331b31d
                                                            0x2e30f07
                                                            0x19cc19f
                                                            0x2e7ac30
                                                            0x2e7ae3f
                                                            0x2eb2df5
                                                            0x2ec141c
                                                            0x2e4caed
                                                            /opt/scylladb/libreloc/libpthread.so.0+0x94e1
                                                            /opt/scylladb/libreloc/libc.so.6+0x1016a2
Jul 10 00:43:20 ip-10-250-2-11 scylla[1804]:  [shard 2] seastar - Exceptional future ignored: seastar::named_semaphore_timed_out (Semaphore timed out: _read_concurrency_sem), backtrace:   0x331b31d
                                                            0x2e30f07
                                                            0x19cc19f
                                                            0x2e7ac30
                                                            0x2e7ae3f
                                                            0x2eb2df5
                                                            0x2ec141c
                                                            0x2e4caed
                                                            /opt/scylladb/libreloc/libpthread.so.0+0x94e1
                                                            /opt/scylladb/libreloc/libc.so.6+0x1016a2
Jul 10 00:43:20 ip-10-250-2-11  scylla[1804]:  [shard 3] seastar - Exceptional future ignored: seastar::named_semaphore_timed_out (Semaphore timed out: _read_concurrency_sem), backtrace:   0x331b31d
                                                            0x2e30f07
                                                            0x19cc19f
                                                            0x2e7ac30
                                                            0x2e7ae3f
                                                            0x2eb2df5
                                                            0x2ec141c
                                                            0x2e4caed
                                                            /opt/scylladb/libreloc/libpthread.so.0+0x94e1
                                                            /opt/scylladb/libreloc/libc.so.6+0x1016a2`

**Comment by gnumoreno** [comment]
@gleb-cloudius @psarna since there are LWT and MVs involved, @vladzcloudius asked me to ping both of you.

**Comment by gnumoreno** [comment]
```
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.1.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.1.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.1.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.0.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - exception during mutation write to 10.250.2.11: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - exception during mutation write to 10.250.2.11: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.0.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - exception during mutation write to 10.250.2.11: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - exception during mutation write to 10.250.2.11: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.0.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - exception during mutation write to 10.250.2.11: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - exception during mutation write to 10.250.2.11: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.0.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.1.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
Jul 10 00:43:23 ip-10-250-2-11.ec2.internal scylla[1804]:  [shard 0] storage_proxy - Failed to apply mutation from 10.250.1.11#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 3->0 semaphore)
```

This is new according to @vladzcloudius

**Comment by gnumoreno** [comment]
The above appear only on the problematic node.

**Comment by psarna** [comment]
Decoded trace from the ignored exceptional future:
```
seastar::report_failed_future(std::__exception_ptr::exception_ptr const&) at ./build/release/seastar/./seastar/src/core/future.cc:101
seastar::future_state_base::~future_state_base() at ././seastar/include/seastar/core/future.hh:335
 (inlined by) seastar::future_state<seastar::lw_shared_ptr<query::result>, cache_temperature>::~future_state() at ././seastar/include/seastar/core/future.hh:427
 (inlined by) seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature>::~future() at ././seastar/include/seastar/core/future.hh:977
 (inlined by) std::_Head_base<1ul, seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature>, false>::~_Head_base() at /usr/include/c++/9/tuple:120
 (inlined by) std::_Tuple_impl<1ul, seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature> >::~_Tuple_impl() at /usr/include/c++/9/tuple:326
 (inlined by) std::_Tuple_impl<0ul, seastar::future<>, seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature> >::~_Tuple_impl() at /usr/include/c++/9/tuple:185
 (inlined by) std::tuple<seastar::future<>, seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature> >::~tuple() at /usr/include/c++/9/tuple:893
 (inlined by) apply at ././seastar/include/seastar/core/apply.hh:36
 (inlined by) apply<service::paxos::paxos_state::prepare(tracing::trace_state_ptr, schema_ptr, const query::read_command&, const partition_key&, utils::UUID, bool, query::digest_algorithm, seastar::lowres_clock::time_point)::<lambda()> mutable::<lambda(service::paxos::paxos_state)>::<lambda(auto:125)>, std::tuple<seastar::future<>, seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature> > > at ././seastar/include/seastar/core/apply.hh:44
 (inlined by) apply<service::paxos::paxos_state::prepare(tracing::trace_state_ptr, schema_ptr, const query::read_command&, const partition_key&, utils::UUID, bool, query::digest_algorithm, seastar::lowres_clock::time_point)::<lambda()> mutable::<lambda(service::paxos::paxos_state)>::<lambda(auto:125)>, std::tuple<seastar::future<>, seastar::future<seastar::lw_shared_ptr<query::result>, cache_temperature> > > at ././seastar/include/seastar/core/future.hh:1664
 (inlined by) operator() at ././seastar/include/seastar/core/future.hh:1208
 (inlined by) run_and_dispose at ././seastar/include/seastar/core/future.hh:522
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2123
seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2538
seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2521
 (inlined by) seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2693
seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}::operator()() const at ./build/release/seastar/./seastar/src/core/reactor.cc:3851
std::function<void ()>::operator()() const at /usr/include/c++/9/bits/std_function.h:688
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:52
```

I'm not sure if it's related to the root cause, but we probably shouldn't ignore an exceptional future here anyway.

**Comment by gleb-cloudius** [comment]
Yes, this is something that should be fixed, but it a symptom of the
problem not the cause. A query to local table failed (probably due to
overload).

--
			Gleb.

**Comment by avikivity** [comment]
@gleb-cloudius we did some work to make sure queries to local tables never fail due to overload.

**Comment by gleb-cloudius** [comment]
On Mon, Jul 13, 2020 at 01:29:06AM -0700, Avi Kivity wrote:
> @gleb-cloudius we did some work to make sure queries to local tables never fail due to overload.
> 
How did we do it? Note that those queries do use timeouts.

--
			Gleb.

**Comment by slivne** [comment]
@avikivity / @gleb-cloudius ?

**Comment by avikivity** [comment]
No, those were only to the schema/gossip tables. Paxos should continue using timeouts.

**Comment by avikivity** [comment]
But what about the failed future? It shouldn't be dropped like that.

**Comment by gleb-cloudius** [comment]
On Thu, Jul 16, 2020 at 02:46:30AM -0700, Avi Kivity wrote:
> But what about the failed future? It shouldn't be dropped like that.
> 
Sure. We need to catch it. I mean that the trace is the symptom, not the
cause of the problem.

--
			Gleb.
